File: src/main/java/com/yerti/core/recipe/CustomRecipe.java
package com.yerti.core.recipe;

import org.bukkit.inventory.ItemStack;
import org.bukkit.inventory.ShapedRecipe;

import java.util.HashMap;
import java.util.Map;

public class CustomRecipe {

    private ShapedRecipe recipe;
    //Output, Matrix Input
    private Map<Character, ItemStack> matrixKey = new HashMap<>();
    private static Map<ItemStack, ItemStack[]> matrixMap = new HashMap<>();


    public CustomRecipe(ItemStack output) {
        recipe = new ShapedRecipe(output);
    }

    public void shape(String... shape) {
        recipe.shape(shape);
    }

    public void setIngredient(char key, ItemStack ingredient) {
        recipe.setIngredient(key, ingredient.getType());
        matrixKey.put(key, ingredient);
    }

    public ShapedRecipe build() {
        ItemStack[] matrix = new ItemStack[9];

        int index = 0;

        for (String string : recipe.getShape()) {
            for (char c : string.toCharArray()) {
                matrix[index] = matrixKey.get(c);
                index++;
            }
        }

        matrixMap.put(recipe.getResult(), matrix);


        return recipe;


    }

    public static Map<ItemStack, ItemStack[]> getMatrixMap() {
        return matrixMap;
    }
}

File: src/main/java/com/yerti/core/recipe/CustomRecipe.kt
package com.yerti.core.recipe

import org.bukkit.inventory.ItemStack
import org.bukkit.inventory.ShapedRecipe
import java.util.*

class CustomRecipe(output: ItemStack?) {
    private val recipe: ShapedRecipe
    //Output, Matrix Input
    private val matrixKey: MutableMap<Char, ItemStack> = HashMap()

    fun shape(vararg shape: String?) {
        recipe.shape(*shape)
    }

    fun setIngredient(key: Char, ingredient: ItemStack) {
        recipe.setIngredient(key, ingredient.type)
        matrixKey[key] = ingredient
    }

    fun build(): ShapedRecipe {
        val matrix = arrayOfNulls<ItemStack>(9)
        var index = 0
        for (string in recipe.shape) {
            for (c in string.toCharArray()) {
                matrix[index] = matrixKey[c]
                index++
            }
        }
        matrixMap[recipe.result] = matrix
        return recipe
    }

    companion object {
        private val matrixMap: MutableMap<ItemStack, Array<ItemStack?>> = HashMap()
        fun getMatrixMap(): Map<ItemStack, Array<ItemStack?>>? {
            return matrixMap
        }
    }

    init {
        recipe = ShapedRecipe(output)
    }
}